<a href="https://colab.research.google.com/github/bhumong/ai-bootcamp/blob/main/project/05_database_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project: Database

**Instructions for Students:**

Please carefully follow these steps to complete and submit your project:

1. **Make a copy of the Project**: Please make a copy of this project either to your own Google Drive or download locally. Work on the copy of the project. The master project is **Read-Only**, meaning you can edit, but it will not be saved when you close the master project. To avoid total loss of your work, remember to make a copy.

2. **Completing the Project**: You are required to work on and complete all tasks in the provided project. Be disciplined and ensure that you thoroughly engage with each task.
   
3. **Creating a Google Drive Folder**: Each of you must create a new folder on your Google Drive. This will be the repository for all your completed project files, aiding you in keeping your work organized and accessible.
   
4. **Uploading Completed Project**: Upon completion of your project, make sure to upload all necessary files, involving codes, reports, and related documents into the created Google Drive folder. Save this link in the 'Student Identity' section and also provide it as the last parameter in the `submit` function that has been provided.
   
5. **Sharing Folder Link**: You're required to share the link to your project Google Drive folder. This is crucial for the submission and evaluation of your project.
   
6. **Setting Permission to Public**: Please make sure your Google Drive folder is set to public. This allows your instructor to access your solutions and assess your work correctly.

Adhering to these procedures will facilitate a smooth project evaluation process for you and the reviewers.

## Project Description

In this project, you'll work with the Kaggle E-commerce dataset. Your tasks are to import this dataset into a SQLite database, explore it using SQL to identify key attributes and patterns, and perform basic data cleaning procedures. This project is designed to improve your skills in database management, SQL usage, and data cleaning.

## Grading Criteria

There are 13 criterias in this project, all have the same weight divided into two: 8 criterias for SQL and 5 criterias for Elasticssearch. Each task will give you either 100 point if you are correct and 0 if you are wrong. The final score for the project will the the average of all 13 tasks.

* Criteria 1, Task I.3: This task will assess your ability to create an SQL database, in this case using sqlite. To ensure correct data is imported, this will check that database table have the correct amount of columns and the correct column names. This will ensure a uniform dataset for all students.

* Criteria 2, Task I.3: Same as Criteria 1, instead of columns, this will check that the total rows of data in the table is correct. This will ensure a uniform dataset for all students.

* Criteria 3, Task I.3.4 query 1: The task will assess your ability to understand and perform basic dataset check, that one row in the database does not guarantee uniqueness for individual columns.

* Criteria 4, Task I.3.4 query 1: Same as Criteria 3.

* Criteria 5, Task I.3.4 query 2: This task will assess your ability to create a SQL query that will get the data, perform arithmetic operations on the data, sort and limit the data returned.

* Criteria 6, Task I.3.4 query 3: Same as Criteria 5.

* Criteria 7, Task I.3.4 query 4: This task will assess your ability to create a SQL query that will get the data, perform arithmetic operations on the data, sort, limit the data and structure the result in a concise format where the all the data are retrieved and presented in a format that is easy to understand can be further processed through programming.

* Criteria 8, Task I.3.4 query 5: Same as Criteria 7.

* Criteria 9, Task II.3: This task will assess your ability to create an Elasticsearch Index. To ensure correct data is imported, this will check that the index have the correct amount of columns the the correct column names. This will ensure a uniform dataset for all students.

* Criteria 10, Task II.3: Same as Criteria 9, instead of columns, this will check that the total rows of data in the table is correct. This will ensure a uniform dataset for all students.

* Criteria 11, Task II.4 query 1: This task will assess your ability to perform a basic elastic search and return the result.

* Criteria 12, Task II.4 query 2: This task will assess your ability to perform an elastic search, perform arithmetic operations on the data and return the result.

* Criteria 13, Task II.4 query 3: This task will assess your ability to perform an elastic search, perform arithmetic operations on the data, sort and limit the data returned.


## Student Identity

In [14]:
# @title #### Student Identity
student_id = "" # @param {type:"string"}
name = "" # @param {type:"string"}
drive_link = ""  # @param {type:"string"}

assignment_id = "00_database_project"

# Import grader package
!pip install rggrader
# from rggrader import submit, submit_image
def submit(*args):
    pass
def submit_image(*args):
    pass


[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


Rmember to make a copy of this notebook in your Google Drive and work in your own copy.

## I. Loading and Querying with SQL

In this task, you will use Kaggle E-commerce dataset. You need to download it from the link here: [Kaggle E-commerce data](https://www.kaggle.com/datasets/carrie1/ecommerce-data?select=data.csv).

After you download the dataset, you will then import the dataset into SQL format, using SQLite.

### I.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [19]:
!pip install sqlalchemy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 3.9 MB/s eta 0:00:00a 0:00:01

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


In [ ]:
#!/bin/bash
!kaggle datasets download carrie1/ecommerce-data

In [1]:
!unzip ecommerce-data.zip

Archive:  ecommerce-data.zip
  inflating: data.csv                


In [ ]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below



### I.2. Data Loading and SQL CREATE and INSERT to database table

Now you can create the database and tables. Follow the same column naming as the dataset.

Next insert all the data from the dataset into the table in the database.

In [15]:
import pandas as pd
df = pd.read_csv('data.csv', encoding='latin-1')
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [ ]:
import sqlite3

# Connect to SQLite database (creates the file if it doesn't exist)
conn = sqlite3.connect('ecommerce.sqlite')
cursor = conn.cursor()

# Create the 'commerce' table if it doesn't exist
cursor.execute('''
CREATE TABLE IF NOT EXISTS commerce (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    InvoiceNo TEXT NOT NULL,
    StockCode TEXT NOT NULL,
    Description TEXT,
    Quantity INTEGER NOT NULL,
    InvoiceDate DATETIME,
    UnitPrice REAL,
    CustomerID INTEGER,
    Country TEXT
)
''')

# Commit the table creation
conn.commit()

In [39]:
df.to_sql("commerce", engine, if_exists='append', index=False)


541909

In [40]:
cursor.execute("SELECT * FROM commerce LIMIT 10;")
rows = cursor.fetchall()
for row in rows:
    print(row)  # Each row is a tuple

(1, '536365', '85123A', 'WHITE HANGING HEART T-LIGHT HOLDER', 6, '12/1/2010 8:26', 2.55, 17850, 'United Kingdom')
(2, '536365', '71053', 'WHITE METAL LANTERN', 6, '12/1/2010 8:26', 3.39, 17850, 'United Kingdom')
(3, '536365', '84406B', 'CREAM CUPID HEARTS COAT HANGER', 8, '12/1/2010 8:26', 2.75, 17850, 'United Kingdom')
(4, '536365', '84029G', 'KNITTED UNION FLAG HOT WATER BOTTLE', 6, '12/1/2010 8:26', 3.39, 17850, 'United Kingdom')
(5, '536365', '84029E', 'RED WOOLLY HOTTIE WHITE HEART.', 6, '12/1/2010 8:26', 3.39, 17850, 'United Kingdom')
(6, '536365', '22752', 'SET 7 BABUSHKA NESTING BOXES', 2, '12/1/2010 8:26', 7.65, 17850, 'United Kingdom')
(7, '536365', '21730', 'GLASS STAR FROSTED T-LIGHT HOLDER', 6, '12/1/2010 8:26', 4.25, 17850, 'United Kingdom')
(8, '536366', '22633', 'HAND WARMER UNION JACK', 6, '12/1/2010 8:28', 1.85, 17850, 'United Kingdom')
(9, '536366', '22632', 'HAND WARMER RED POLKA DOT', 6, '12/1/2010 8:28', 1.85, 17850, 'United Kingdom')
(10, '536367', '84879', 'ASSO

In [42]:
import sqlite3

def execute_query(sql, parameters=()):
    """Execute a query and return all rows"""
    with sqlite3.connect('ecommerce.sqlite') as conn:
        cursor = conn.cursor()
        cursor.execute(sql, parameters)
        return cursor.fetchall()

def execute_count_query(sql, parameters=()):
    """Execute a query that returns a count value"""
    with sqlite3.connect('ecommerce.sqlite') as conn:
        cursor = conn.cursor()
        cursor.execute(sql, parameters)
        result = cursor.fetchone()
        return result[0] if result else 0

### I.3. Verifying the database

Before we go to the next section, let's check our database first. You need to change the db_name and table_name with your database name and table name, afterwards you can just run the code block. The result should be:

- Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
- Number of rows: 541909

In [41]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('ecommerce.sqlite') #change db_name to your database name.
cur = conn.cursor()

# Get column names
cur.execute("PRAGMA table_info(commerce)") #change commerce to your table name
sql_columns = [column[1] for column in cur.fetchall()] #the variable sql_columns is used for submission, do not change the variable name
print(f"Columns: {sql_columns}")

# Get total number of rows
cur.execute("SELECT COUNT(*) FROM commerce") #change commerce to your table name
sql_num_rows = cur.fetchone()[0] #the variable sql_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {sql_num_rows}")

# Close the connection
conn.close()

Columns: ['id', 'InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
Number of rows: 541909


### I.3.4. Running SQL Queries

Once we have the database in SQL format, let's do some data exploration and analysis.

You are part of the engineering team at a global e-commerce company. Your company has a vast catalogue of products and serves customers around the globe. As the company continues to grow, the management decides it's time to expand the business even more.

As an essential part of this expansion, your team is tasked with diving into your company's rich e-commerce dataset to extract crucial insights that will guide the expansion strategy.

1. **Total number of unique products and unique customers:**

   Your first task is to gauge the breadth of your operations. By determining the total number of unique products you sell and the total number of unique customers you serve, you can assess the scale and diversity of your business.

2. **Total revenue for each product:**

   Next, you're going to identify the star performers in your product portfolio. Calculating the total revenue for each product will reveal which items are the biggest revenue drivers. These insights can help guide decisions about product focus and marketing efforts. Limit the result to the top 5 product with the biggest revenue.

3. **Most profitable countries (by total sales):**

   To pinpoint the most fruitful geographic areas for your business, you'll identify which countries generate the most profit. This data will help you understand where your operations are already strong and where there might be potential for regional expansion. Limit the result to the top 5 countries with the biggest profit.

4. **Top-selling products (by total sales) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by sales for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

4. **Top-selling products (by quantity) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by quantity for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

By answering these queries, your team will provide valuable data-driven insights that will directly contribute to the strategic decisions about where and how your business expands. As part of the engineering team, you're not just maintaining the technical infrastructure - you're shaping the future of the company.

> Hint: The Product name is stored in column Descrioption

In [ ]:
# Write your SQL Query here

#use the following variable name to store the result from your SQL Query
sql_num_products = 0
sql_num_customers = 0
sql_top_product_revenues = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_countries = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_selling_products_by_country_by_sales = {} #dictionary where key is country name, value is a list of top-selling products by sales, see example below
sql_top_selling_products_by_country_by_quantity = {} #dictionary where key is country name, value is a list of top-selling products by quantity, see example below


# 1. Find the total number of unique products and unique customers
# Example: sql_num_products = 1559
# Example: sql_num_customers = 1675



[(4070,)]

In [ ]:
sql_num_products = execute_count_query("SELECT count(distinct StockCode) from commerce;")
sql_num_products

4070

In [46]:
sql_num_customers = execute_count_query("SELECT count(distinct CustomerID) from commerce;")
sql_num_customers

4372

In [50]:
# 2. Calculate the total revenue for each product, limited to top 5
# Example: sql_top_product_revenues = [('PRODUCT 1', 149385.12), ('PRODUCT 2', 54895.53), ('PRODUCT 3', 44545.55), ('PRODUCT 4', 38000.00), ('PRODUCT 5', 36000.00)]

sql_top_product_revenues = execute_query("SELECT StockCode, SUM(UnitPrice) AS TotalUnitPrice FROM commerce GROUP BY StockCode ORDER BY TotalUnitPrice DESC LIMIT 5;")
sql_top_product_revenues


[('AMAZONFEE', 249042.68),
 ('M', 214448.41),
 ('DOT', 206252.06),
 ('POST', 46388.47),
 ('22423', 30360.61)]

In [51]:
# 3. Identify the most profitable countries (by total sales), limited to top 5
# Example: sql_top_countries = [('COUNTRY 1', 600000.00), ('COUNTRY 2', 334857.45), ('COUNTRY 3', 245879.00), ('COUNTRY 4', 180000.00), ('COUNTRY 5', 164389.45)]

sql_top_countries = execute_query("SELECT Country, SUM(UnitPrice) AS TotalUnitPrice FROM commerce GROUP BY Country ORDER BY TotalUnitPrice DESC LIMIT 5;")
sql_top_countries


[('United Kingdom', 2245715.474),
 ('EIRE', 48447.19),
 ('France', 43031.99),
 ('Germany', 37666.0),
 ('Singapore', 25108.89)]

In [54]:
# 4. Find the top-selling products (by total sales) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_sales = {
#    'COUNTRY 1': [('PRODUCT 1', 200000.00), ('PRODUCT 2', 100000.00), ('PRODUCT 3', 95348.00), ('PRODUCT 4', 90000.43), ('PRODUCT 5', 80000.00)],
#    'COUNTRY 2': [('PRODUCT 1', 9000.00), ('PRODUCT 2', 7345.00), ('PRODUCT 3', 6934.23), ('PRODUCT 4', 6800.23), ('PRODUCT 5', 4000.00)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('PRODUCT 1', 7442.84), ('PRODUCT 2', 5175.00), ('PRODUCT 3', 3098.0), ('PRODUCT 4', 3014.34), ('PRODUCT 5', 2857.83)],

def get_top_products_by_country():
    query = """
    WITH SalesCTE AS (
        SELECT 
            Country,
            StockCode,
            SUM(UnitPrice * Quantity) AS TotalSales,
            ROW_NUMBER() OVER (PARTITION BY Country ORDER BY SUM(UnitPrice * Quantity) DESC) AS Rank
        FROM commerce
        GROUP BY Country, StockCode
    )
    SELECT Country, StockCode, TotalSales
    FROM SalesCTE
    WHERE Rank <= 5
    ORDER BY Country, Rank;
    """
    
    results = execute_query(query)
    
    output = {}
    for row in results:
        country = row[0]
        product = row[1]
        sales = round(row[2], 2)
        
        if country not in output:
            output[country] = []
        output[country].append((product, sales))
    
    return output

sql_top_selling_products_by_country_by_sales = get_top_products_by_country()
sql_top_selling_products_by_country_by_sales


{'Australia': [('23084', 3375.84),
  ('22722', 2082.0),
  ('21731', 1987.2),
  ('22720', 1978.25),
  ('22423', 1978.2)],
 'Austria': [('POST', 1456.0),
  ('22584', 302.4),
  ('22582', 302.4),
  ('20679', 214.2),
  ('15056N', 214.2)],
 'Bahrain': [('23076', 120.0),
  ('23077', 75.0),
  ('22890', 59.7),
  ('22649', 39.6),
  ('22693', 30.0)],
 'Belgium': [('POST', 4248.0),
  ('22326', 1181.4),
  ('22630', 641.85),
  ('22629', 641.85),
  ('22423', 599.25)],
 'Brazil': [('22423', 175.2),
  ('22722', 82.8),
  ('21430', 81.36),
  ('22366', 67.5),
  ('22699', 61.2)],
 'Canada': [('POST', 550.94),
  ('37370', 534.24),
  ('20727', 82.5),
  ('84077', 60.48),
  ('22383', 49.5)],
 'Channel Islands': [('22423', 517.8),
  ('85099B', 460.7),
  ('22720', 408.0),
  ('85099C', 399.6),
  ('23199', 399.6)],
 'Cyprus': [('22827', 580.0),
  ('85123A', 386.4),
  ('22423', 382.5),
  ('22656', 250.0),
  ('22655', 250.0)],
 'Czech Republic': [('22326', 70.8),
  ('84347', 61.2),
  ('21428', 51.0),
  ('47594B', 46

In [55]:
# 5. Find the top-selling products (by quantity) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_quantity = {
#    'COUNTRY 1': [('COUNTRY 1, PRODUCT 1', 22343), ('COUNTRY 1, PRODUCT 2', 12345), ('COUNTRY 1, PRODUCT 3', 9534), ('COUNTRY 1, PRODUCT 4', 9000), ('COUNTRY 1, PRODUCT 5', 8000)],
#    'COUNTRY 2': [('COUNTRY 2, PRODUCT 1', 9000), ('COUNTRY 2, PRODUCT 2', 7345), ('COUNTRY 2, PRODUCT 3', 6934), ('COUNTRY 2, PRODUCT 4', 6800), ('COUNTRY 2, PRODUCT 5', 4000)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5, PRODUCT 1', 48345), ('COUNTRY 5, PRODUCT 2', 43965), ('COUNTRY 5, PRODUCT 3', 43556), ('COUNTRY 5, PRODUCT 4', 34567), ('COUNTRY 5, PRODUCT 5', 33436)],


def get_top_products_by_country_quantity():
    query = """
    WITH QuantityCTE AS (
        SELECT 
            Country,
            StockCode,
            Country || ', ' || MAX(Description) AS ProductName,
            SUM(Quantity) AS TotalQuantity,
            ROW_NUMBER() OVER (
                PARTITION BY Country 
                ORDER BY SUM(Quantity) DESC
            ) AS Rank
        FROM commerce
        GROUP BY Country, StockCode
    )
    SELECT Country, ProductName, TotalQuantity
    FROM QuantityCTE
    WHERE Rank <= 5
    ORDER BY Country, Rank;
    """
    
    results = execute_query(query)
    
    # Format results into dictionary structure
    output = {}
    for row in results:
        country = row[0]
        product = row[1]
        quantity = row[2]
        
        if country not in output:
            output[country] = []
        output[country].append((product, quantity))
    
    return output

sql_top_selling_products_by_country_by_quantity = get_top_products_by_country_quantity()
sql_top_selling_products_by_country_by_quantity

{'Australia': [('Australia, MINI PAINT SET VINTAGE ', 2916),
  ('Australia, RABBIT NIGHT LIGHT', 1884),
  ('Australia, RED  HARMONICA IN BOX ', 1704),
  ('Australia, RED TOADSTOOL LED NIGHT LIGHT', 1344),
  ('Australia, DOLLY GIRL LUNCH BOX', 1024)],
 'Austria': [('Austria, SET 12 KIDS COLOUR  CHALK STICKS', 288),
  ('Austria, MINI JIGSAW PURDEY', 240),
  ('Austria, PACK OF 6 PANNETONE GIFT BOXES', 144),
  ('Austria, PACK OF 6 SWEETIE GIFT BOXES', 144),
  ('Austria, PLASTERS IN TIN CIRCUS PARADE ', 48)],
 'Bahrain': [('Bahrain, ICE CREAM SUNDAE LIP GLOSS', 96),
  ('Bahrain, DOUGHNUT LIP GLOSS ', 60),
  ('Bahrain, GROW A FLYTRAP OR SUNFLOWER IN TIN', 24),
  ('Bahrain, S/4 PINK FLOWER CANDLES IN BOWL', 12),
  ('Bahrain, MINI CAKE STAND WITH HANGING CAKES', 8)],
 'Belgium': [('Belgium, PACK OF 72 RETROSPOT CAKE CASES', 480),
  ('Belgium, ROUND SNACK BOXES SET OF4 WOODLAND ', 420),
  ('Belgium, PACK OF 60 SPACEBOY CAKE CASES', 384),
  ('Belgium, DOLLY GIRL LUNCH BOX', 339),
  ('Belgium, SP

## II. Indexing and Querying with Elasticsearch

You should have a better understanding of how to use SQL, now let's leverage the power of Elasticsearch to explore and analyze the data. We'll use the same dataset.


### II.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [56]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below
!pip install elasticsearch


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 718.0/718.0 kB 997.2 kB/s eta 0:00:00 0:00:01

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


### II.2. Set up Elasticsearch Index

The first step you need to do which is **preprocessing, let's keep it simple and simply drop all row that have missing values.**

Next, you need to set up an Elasticsearch index. An Elasticsearch index is like a database in a traditional relational database. An index lets you store, search, and analyze big volumes of data quickly and in near real time.

You'll also need to setup the mappings for the data types, let's do that here.

In [68]:
# Preprocessing here

# Load your dataset and create an Elasticsearch index
# Hints: use the BulkAPI for faster creation of the index
from elasticsearch import Elasticsearch

df = pd.read_csv('data.csv', encoding='latin-1' ,parse_dates=['InvoiceDate'])
df.dropna(inplace=True)
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[406829 rows x 8 columns]

In [107]:
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

index_name = "commerce"

mappings = {
    "mappings": {
        "properties": {
            "InvoiceNo": {"type": "keyword"},
            "StockCode": {"type": "keyword"},
            "Description": { # Modified Description mapping
                "type": "text",
                "fields": {
                    "keyword": {"type": "keyword"} # Add keyword sub-field
                }
            },
            "Quantity": {"type": "integer"},
            "InvoiceDate": {"type": "date"},
            "UnitPrice": {"type": "float"},
            "CustomerID": {"type": "keyword"},
            "Country": {"type": "keyword"}
        }
    }
}

if es.indices.exists(index=index_name):
    es.indices.delete(index=index_name)

es.indices.create(index=index_name, body=mappings)

ObjectApiResponse({'acknowledged': True, 'shards_acknowledged': True, 'index': 'commerce'})

In [108]:
from elasticsearch.helpers import bulk

documents = df.to_dict(orient='records')
def generate_actions(docs):
    for doc in docs:
        yield {
            "_index": index_name,
            "_source": doc
        }

In [109]:
from elasticsearch.helpers import bulk

try:
    success, failed = bulk(
        es,
        generate_actions(documents),
        raise_on_error=False  # Don't raise exceptions, return errors instead
    )
    print(f"Successfully inserted {success} documents")
    
    # Inspect failed documents
    if failed:
        print("Failed documents:")
        for error in failed:
            print(f"Error: {error}")

except Exception as e:
    print(f"Unexpected error: {e}")

Successfully inserted 406829 documents


In [110]:
my_index_name = index_name


### II.3. Verifying the index

Before we go to the next section, let's check our index first. You need to change the my_index_name with your chosen index name, afterwards you can just run the code block. The result should be:

- Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
- Number of rows: 406829

In [111]:
from elasticsearch import Elasticsearch

# Connect to the Elasticsearch server
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

index_name = my_index_name #change the my_index_name value

# Get index mapping (equivalent to getting column names in SQL)
mapping = es.indices.get_mapping(index=index_name)
es_columns = list(mapping[index_name]['mappings']['properties'].keys()) #the variable es_columns is used for submission, do not change the variable name
print(f"Columns: {es_columns}")

# Count documents in the index (equivalent to counting rows in SQL)
es_num_rows = es.count(index=index_name)['count'] #the variable es_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {es_num_rows}")


Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
Number of rows: 406829


### II.4. Running Elasticsearch queries

Once your e-commerce data is cleaned and ready, let's analyse the data to gain insights that will drive the expansion strategy, this time using Elasticsearch and we'll focus on one specific country, which is Germany.

1. **Find all transactions for a specific country**

   Your company operates globally, but you want to understand better how different countries contribute to your sales. By finding all transactions that happened in a specific country, you can get a clearer picture of your company's reach and performance in that location. In our case, we'll focus on the country Germany.


2. **Find all unique products in Germany**

   Your first task is to understand the breadth of your company's product portfolio in Germany. By determining the total number of unique products you sell, you can assess the scale and diversity of your operations.


3. **Find the top 5 most purchased products in Germany**

   Next up, you'll identify the best-selling products in your portfolio in Germany. Knowing which items are the most purchased can help guide decisions about product focus and marketing efforts. We'll limit the result to 5 most purchased products.


By executing these tasks, you will enable your team to provide data-driven insights which will directly contribute to the strategic decisions about your business expansion. As part of the data engineering team, you're not only maintaining the technical infrastructure - you're shaping the future of the company.


In [112]:
# Write your Elasticsearch Query here

#use the following variable name to store the result from your Elasticsearch Query
es_total_unique_products = 0
es_transactions_in_germany = 0
es_top_products = {} #dictionary where key is product name, value is the quantity sold, see example below

# 1. Find all transactions for a specific country (Germany)
# Example: es_transactions_in_germany = 7584




In [113]:
# 1. Find all transactions for a specific country (Germany)
# Example: es_transactions_in_germany = 7584

query = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    }
}

response = es.search(index="commerce", body=query)

total_transactions = response["hits"]["total"]["value"]
print(f"Total transactions in Germany: {total_transactions}")

for hit in response["hits"]["hits"][:5]:
    print(hit["_source"])

es_transactions_in_germany = total_transactions

Total transactions in Germany: 9495
{'InvoiceNo': '540553', 'StockCode': '22858', 'Description': 'EASTER TIN KEEPSAKE', 'Quantity': 48, 'InvoiceDate': '2011-01-10T09:48:00', 'UnitPrice': 1.65, 'CustomerID': 12500.0, 'Country': 'Germany'}
{'InvoiceNo': '540553', 'StockCode': '22861', 'Description': 'EASTER TIN CHICKS IN GARDEN', 'Quantity': 48, 'InvoiceDate': '2011-01-10T09:48:00', 'UnitPrice': 1.65, 'CustomerID': 12500.0, 'Country': 'Germany'}
{'InvoiceNo': '540553', 'StockCode': '22585', 'Description': 'PACK OF 6 BIRDY GIFT TAGS', 'Quantity': 144, 'InvoiceDate': '2011-01-10T09:48:00', 'UnitPrice': 1.06, 'CustomerID': 12500.0, 'Country': 'Germany'}
{'InvoiceNo': '540553', 'StockCode': '22366', 'Description': 'DOORMAT AIRMAIL ', 'Quantity': 2, 'InvoiceDate': '2011-01-10T09:48:00', 'UnitPrice': 7.95, 'CustomerID': 12500.0, 'Country': 'Germany'}
{'InvoiceNo': '540553', 'StockCode': 'POST', 'Description': 'POSTAGE', 'Quantity': 1, 'InvoiceDate': '2011-01-10T09:48:00', 'UnitPrice': 18.0, 'C

In [114]:
# 2. Find all unique products in Germany
# Hints: Use set() to remove duplicate values
# Example: es_total_unique_products = 45345
query = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    },
    "aggs": {
        "unique_products": {
            "terms": {
                "field": "StockCode",
                "size": 10000
            }
        }
    },
    "size": 0
}

# Execute the search
response = es.search(index="commerce", body=query)

# Extract the result
total_unique_products = response["aggregations"]["unique_products"]["buckets"]
es_total_unique_products = len(total_unique_products)
es_total_unique_products



1671

In [115]:
query = {
    "query": {
        "term": {
            "Country": "Germany"
        }
    },
    "aggs": {
        "top_products": {
            "terms": {
                "field": "Description.keyword", # Now Description.keyword should work!
                "size": 5,
                "order": { "total_quantity": "desc" }
            },
            "aggs": {
                "total_quantity": {
                    "sum": { "field": "Quantity" }
                }
            }
        }
    },
    "size": 0
}

response = es.search(index="commerce", body=query)

# Extract and format results
buckets = response["aggregations"]["top_products"]["buckets"]
es_top_products = {bucket["key"]: int(bucket["total_quantity"]["value"]) for bucket in buckets}

print(f"Top 5 products in Germany: {es_top_products}")

Top 5 products in Germany: {'ROUND SNACK BOXES SET OF4 WOODLAND ': 1218, 'ASSORTED COLOURS SILK FAN': 1164, 'POSTAGE': 1104, 'WOODLAND CHARLOTTE BAG': 1019, 'PACK OF 72 RETROSPOT CAKE CASES': 1002}


In [118]:
from rggrader import submit, submit_image

## Submission

Once you are satisfied with the performance of your model, then you run the code block below to submit your project.


In [119]:
# Submit Method

# Do not change the code below
question_id = "01_sql_columns"
submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
question_id = "02_sql_num_rows"
submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
question_id = "03_sql_num_products"
submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
question_id = "04_sql_num_customers"
submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
question_id = "05_sql_top_product_revenues"
submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
question_id = "06_sql_top_countries"
submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
question_id = "07_sql_top_selling_products_by_country_by_sales"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
question_id = "08_sql_top_selling_products_by_country_by_quantity"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)

question_id = "09_es_columns"
submit(student_id, name, assignment_id, str(es_columns), question_id, drive_link)
question_id = "10_es_num_rows"
submit(student_id, name, assignment_id, str(es_num_rows), question_id, drive_link)
question_id = "11_es_total_unique_products"
submit(student_id, name, assignment_id, str(es_total_unique_products), question_id, drive_link)
question_id = "12_es_transactions_in_germany"
submit(student_id, name, assignment_id, str(es_transactions_in_germany), question_id, drive_link)
question_id = "13_es_top_products"
submit(student_id, name, assignment_id, str(es_top_products), question_id, drive_link)

'Assignment successfully submitted'

## FIN